// prisma-db-introspection/dist/common/normalization/adjustJoinTableCardinality.js
"use strict";
var __values = (this && this.__values) || function (o) {
    var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
    if (m) return m.call(o);
    return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
};
Object.defineProperty(exports, "__esModule", { value: true });
var prisma_datamodel_1 = require("prisma-datamodel");
var AdjustJoinTableCardinality = /** @class */ (function () {
    function AdjustJoinTableCardinality(baseModel) {
        this.baseModel = baseModel;
    }
    AdjustJoinTableCardinality.prototype.normalizeType = function (type, ref) {
        var e_1, _a;
        var _loop_1 = function (field) {
            // Only look at n:n relations.
            if (typeof field.type !== 'string' && field.isList) {
                // TODO: Subclass
                // Fid the reference field.
                var refField = ref.fields.find(function (x) { return x.name === field.name; });
                if (refField === undefined || typeof refField.type === 'string')
                    return "continue";
                if (refField.type.name !== field.type.name)
                    return "continue";
                // If the reference is not a list. We restrict it to a non-list and add
                // a link: TABLE directive (compatability mode).
                if (!refField.isList) {
                    field.isList = false;
                    field.isRequired = refField.isRequired;
                    field.directives.push(this_1.createLinkTableDirective());
                    if (field.relatedField !== null) {
                        field.relatedField.directives.push(this_1.createLinkTableDirective());
                    }
                }
            }
        };
        var this_1 = this;
        try {
            for (var _b = __values(type.fields), _c = _b.next(); !_c.done; _c = _b.next()) {
                var field = _c.value;
                _loop_1(field);
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
    };
    AdjustJoinTableCardinality.prototype.normalize = function (model) {
        var e_2, _a;
        var _loop_2 = function (type) {
            // TODO: We should move all tooling for finding types or fields into some common class.
            var ref = this_2.baseModel.types.find(function (x) { return x.name === type.name || x.databaseName === type.name; });
            if (ref !== undefined) {
                this_2.normalizeType(type, ref);
            }
        };
        var this_2 = this;
        try {
            for (var _b = __values(model.types), _c = _b.next(); !_c.done; _c = _b.next()) {
                var type = _c.value;
                _loop_2(type);
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_2) throw e_2.error; }
        }
    };
    AdjustJoinTableCardinality.prototype.createLinkTableDirective = function () {
        return {
            name: prisma_datamodel_1.DirectiveKeys.relation,
            arguments: {
                link: 'TABLE',
            },
        };
    };
    return AdjustJoinTableCardinality;
}());
exports.AdjustJoinTableCardinality = AdjustJoinTableCardinality;
